docs/user/exercises/programming-exercise-AI-proof-guidelines.inc

```diff
@@ -0,0 +1,491 @@
+   types, such as custom classes and data structures, often involve
+   additional levels of complexity and variation in their
+   implementation.
+
+   Here are some of the complex data types that can improve the AI
+   security of the exercises:
+
+   -  Duration
+
+   -  BigInteger, BigDecimal
+
+   -  LocalDate
+
+   -  LocalTime
+
+   -  Locale
+
+   -  StringBuilder
+
+   -  Currency
+
+   -  Custom classes implemented in the template repository
+
+   In the following example you can see how ChatGPT tries to work around
+   using BigInteger. In the exercise the students needed to
+   validate the IBAN that was given as a parameter with a checkSum by using methods from the
+   BigInteger library. ChatGPT on the other hand first converted the BIGInteger into a String
+   to perform the required operations.
+
+   Generated solution:
+
+   .. code:: java
+
+      public static boolean validateIBAN(BigInteger iban) {
+          String rearranged = iban.toString().substring(4) +
+               iban.toString().substring(0, 4);
+
+          StringBuilder numericString = new StringBuilder();
+          for (int i = 0; i < rearranged.length(); i++) {
+            char c = rearranged.charAt(i);
+            if (Character.isLetter(c)) {
+              numericString.append((int)c - 55);
+            } else {
+              numericString.append(c);
+            }
+          }
+
+           BigInteger numeric = new BigInteger(numericString.toString());
+          return numeric.mod(BigInteger.valueOf(97)).equals(BigInteger.ONE);
+      }
+
+   Example solution:
+
+   .. code:: java
+
+      public static boolean validateIBAN(BigInteger iban) {
+
+          BigInteger checkSum = iban.divide(BigInteger.TEN.pow(18));
+          BigInteger accountNumber = iban.mod(BigInteger.TEN.pow(18));
+          BigInteger shifted = accountNumber.multiply(BigInteger.TEN.pow(6));
+          BigInteger deInserted = shifted.add(new BigInteger("131400"));
+          BigInteger deCheckSumInserted = deInserted.add(checkSum);
+          BigInteger remainder = deCheckSumInserted.mod(new BigInteger("97"));
+          return remainder.intValue() == 1;
+      }
+
+   .. note::
+      Give information to the students on which datatypes they should use. You can either do that
+      by stating it in the test cases or directly in the problem statement.
+      Do not overstate it in the problem statement, since ChatGPT will then also have that information.
+
+-  **Standard Java methods:** Make use of letting students implement
+   specific parts of methods on their own instead of using standard
+   functionality provided by Java. AI often takes advantage of these to
+   save on implementation. It should be explicitly inside of a method
+   that has additional functionality instead of just on its own. It
+   obviously needs to be checked for these standard methods then in the
+   test cases. Some functionality that can be implemented from scratch.
```

Comment:
From my experience ChatGPT is really really good at writing code for doing this stuff manually.
While developing Iris it is almost impossible to prevent it from generating such code because of how common it is in its dataset.
With even halfway decent prompting it would be extremely easy to get ChatGPT to do even more complex stuff.

Reply:
Another problem is that students "waste" time with this. If I want to test their knowledge about Threads or any other topic, writing these simple functions will just take up time that steals time from the actual content of the exercise.

Reply:
It makes more sense to apply this in the right context. Obviously when they are at the stage of implementing more complex functionality, they should be able to take the standard methods. When they just covered sorting in the lecture and need to implement it in a method, then we need to prevent the use of standard methods. This is probably more a general guideline and not specifically for AI proof exercises. I will write that as a note afterwards.